consensus/state_processing/src/per_block_processing/process_operations.rs

```diff
@@ -129,13 +135,110 @@ pub mod altair_deneb {
 
         let attesting_indices = &verify_attestation_for_block_inclusion(
             state,
-            attestation,
+            &Attestation::Base(attestation.clone()),
             ctxt,
             verify_signatures,
             spec,
         )
         .map_err(|e| e.into_with_index(att_index))?
-        .attesting_indices;
+        .attesting_indices();
+
+        // Matching roots, participation flag indices
+        let data = &attestation.data;
+        let inclusion_delay = state.slot().safe_sub(data.slot)?.as_u64();
+        let participation_flag_indices =
+            get_attestation_participation_flag_indices(state, data, inclusion_delay, spec)?;
+
+        // Update epoch participation flags.
+        let total_active_balance = state.get_total_active_balance()?;
+        let base_reward_per_increment = BaseRewardPerIncrement::new(total_active_balance, spec)?;
+        let mut proposer_reward_numerator = 0;
+        for index in attesting_indices {
+            let index = *index as usize;
+
+            for (flag_index, &weight) in PARTICIPATION_FLAG_WEIGHTS.iter().enumerate() {
+                let epoch_participation = state.get_epoch_participation_mut(data.target.epoch)?;
+                let validator_participation = epoch_participation
+                    .get_mut(index)
+                    .ok_or(BeaconStateError::ParticipationOutOfBounds(index))?;
+
+                if participation_flag_indices.contains(&flag_index)
+                    && !validator_participation.has_flag(flag_index)?
+                {
+                    validator_participation.add_flag(flag_index)?;
+                    proposer_reward_numerator.safe_add_assign(
+                        get_base_reward(state, index, base_reward_per_increment, spec)?
+                            .safe_mul(weight)?,
+                    )?;
+
+                    if flag_index == TIMELY_TARGET_FLAG_INDEX {
+                        update_progressive_balances_on_attestation(
+                            state,
+                            data.target.epoch,
+                            index,
+                        )?;
+                    }
+                }
+            }
+        }
+
+        let proposer_reward_denominator = WEIGHT_DENOMINATOR
+            .safe_sub(PROPOSER_WEIGHT)?
+            .safe_mul(WEIGHT_DENOMINATOR)?
+            .safe_div(PROPOSER_WEIGHT)?;
+        let proposer_reward = proposer_reward_numerator.safe_div(proposer_reward_denominator)?;
+        increase_balance(state, proposer_index as usize, proposer_reward)?;
+        Ok(())
+    }
+}
+
+pub mod electra {
```

Comment:
Not necessary to duplicate all this logic. The only difference is when converting an on-chain attestation to an indexed attestation which can be common in all forks. Just convert attestations from any fork into the electra indexed attestation. The relevant attestation data for reward processing is the same too.